File: backend/scripts/perf_smoke.py
"""Lightweight performance smoke checks for API latency budgets.

Writes a JSON report consumable by CI artifact uploads.
"""

from __future__ import annotations

import json
import os
import statistics
import time
from dataclasses import asdict, dataclass
from pathlib import Path

import httpx


@dataclass
class EndpointStats:
    endpoint: str
    samples_ms: list[float]
    avg_ms: float
    p95_ms: float
    max_ms: float


@dataclass
class PerfReport:
    base_url: str
    sample_count: int
    budget_avg_ms: float
    budget_p95_ms: float
    endpoints: list[EndpointStats]


def percentile(values: list[float], pct: float) -> float:
    if not values:
        return 0.0
    ordered = sorted(values)
    idx = min(len(ordered) - 1, max(0, int(round((pct / 100) * (len(ordered) - 1)))))
    return ordered[idx]


def collect_endpoint(client: httpx.Client, endpoint: str, sample_count: int) -> EndpointStats:
    latencies: list[float] = []
    for _ in range(sample_count):
        start = time.perf_counter()
        response = client.get(endpoint)
        elapsed_ms = (time.perf_counter() - start) * 1000
        if response.status_code >= 400:
            raise RuntimeError(f"{endpoint} returned {response.status_code}: {response.text[:200]}")
        latencies.append(elapsed_ms)
    return EndpointStats(
        endpoint=endpoint,
        samples_ms=[round(item, 2) for item in latencies],
        avg_ms=round(statistics.fmean(latencies), 2),
        p95_ms=round(percentile(latencies, 95), 2),
        max_ms=round(max(latencies), 2),
    )


def main() -> int:
    base_url = os.getenv("PERF_API_BASE_URL", "http://127.0.0.1:8000").rstrip("/")
    sample_count = int(os.getenv("PERF_SAMPLE_COUNT", "8"))
    budget_avg_ms = float(os.getenv("PERF_BUDGET_AVG_MS", "800"))
    budget_p95_ms = float(os.getenv("PERF_BUDGET_P95_MS", "1500"))
    report_path = Path(os.getenv("PERF_REPORT_PATH", "backend/perf-results/api-latency-smoke.json"))

    endpoints = [
        "/api/health/live",
        "/api/health/ready",
        "/api/system/analytics",
    ]

    stats: list[EndpointStats] = []
    with httpx.Client(base_url=base_url, timeout=20.0) as client:
        for endpoint in endpoints:
            stats.append(collect_endpoint(client, endpoint, sample_count))

    report = PerfReport(
        base_url=base_url,
        sample_count=sample_count,
        budget_avg_ms=budget_avg_ms,
        budget_p95_ms=budget_p95_ms,
        endpoints=stats,
    )

    report_path.parent.mkdir(parents=True, exist_ok=True)
    report_path.write_text(
        json.dumps(
            {
                "base_url": report.base_url,
                "sample_count": report.sample_count,
                "budget_avg_ms": report.budget_avg_ms,
                "budget_p95_ms": report.budget_p95_ms,
                "endpoints": [asdict(item) for item in report.endpoints],
            },
            indent=2,
        )
    )

    failures: list[str] = []
    for item in stats:
        if item.avg_ms > budget_avg_ms:
            failures.append(f"{item.endpoint} avg {item.avg_ms}ms > budget {budget_avg_ms}ms")
        if item.p95_ms > budget_p95_ms:
            failures.append(f"{item.endpoint} p95 {item.p95_ms}ms > budget {budget_p95_ms}ms")

    if failures:
        print("Performance smoke failed:")
        for message in failures:
            print(f"- {message}")
        return 1

    print(f"Performance smoke passed. Report: {report_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend/scripts/perf_smoke.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { performance } from "node:perf_hooks";

const baseUrl = (process.env.PERF_FRONTEND_BASE_URL ?? "http://127.0.0.1:3000").replace(/\/$/, "");
const samples = Number(process.env.PERF_FRONTEND_SAMPLE_COUNT ?? 5);
const reportPath = process.env.PERF_FRONTEND_REPORT_PATH ?? "frontend/perf-results/frontend-latency-smoke.json";
const avgBudget = Number(process.env.PERF_FRONTEND_AVG_BUDGET_MS ?? 1200);
const p95Budget = Number(process.env.PERF_FRONTEND_P95_BUDGET_MS ?? 2500);

const routes = ["/", "/dashboard", "/schedule"];

function percentile(values, pct) {
  const ordered = [...values].sort((a, b) => a - b);
  const idx = Math.min(ordered.length - 1, Math.max(0, Math.round(((pct / 100) * (ordered.length - 1)))));
  return ordered[idx] ?? 0;
}

async function measureRoute(route) {
  const latencies = [];
  for (let i = 0; i < samples; i += 1) {
    const start = performance.now();
    const response = await fetch(`${baseUrl}${route}`);
    const elapsed = performance.now() - start;
    if (!response.ok && response.status !== 404) {
      throw new Error(`Unexpected status ${response.status} for ${route}`);
    }
    latencies.push(Number(elapsed.toFixed(2)));
  }

  const avg = Number((latencies.reduce((sum, value) => sum + value, 0) / latencies.length).toFixed(2));
  const p95 = Number(percentile(latencies, 95).toFixed(2));
  const max = Number(Math.max(...latencies).toFixed(2));

  return {
    route,
    samplesMs: latencies,
    avgMs: avg,
    p95Ms: p95,
    maxMs: max,
  };
}

async function main() {
  const endpointReports = [];
  for (const route of routes) {
    endpointReports.push(await measureRoute(route));
  }

  const output = {
    baseUrl,
    samples,
    avgBudgetMs: avgBudget,
    p95BudgetMs: p95Budget,
    routes: endpointReports,
  };

  await fs.mkdir(path.dirname(reportPath), { recursive: true });
  await fs.writeFile(reportPath, `${JSON.stringify(output, null, 2)}\n`, "utf8");

  const failures = [];
  for (const route of endpointReports) {
    if (route.avgMs > avgBudget) {
      failures.push(`${route.route} avg ${route.avgMs}ms > ${avgBudget}ms`);
    }
    if (route.p95Ms > p95Budget) {
      failures.push(`${route.route} p95 ${route.p95Ms}ms > ${p95Budget}ms`);
    }
  }

  if (failures.length > 0) {
    console.error("Frontend performance smoke failed:");
    failures.forEach((failure) => console.error(`- ${failure}`));
    process.exitCode = 1;
    return;
  }

  console.log(`Frontend performance smoke passed. Report: ${reportPath}`);
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
